### PR TITLE
[jit] Resolve Python values before compiler builtins

### PR DESCRIPTION
**Pull request by driazati**
Stacked PRs
 * #34112 - [jit] Support custom error messages for `raise`
 * **#34910 - [jit] Resolve Python values before compiler builtins**

See #34112 for context, TLDR: This PR fixes resolution so that Python values are preferred over the compiler builtins. As a follow up some of the compiler globals can be removed and bound at the Python level.

Differential Revision: [D20496942](https://our.internmc.facebook.com/intern/diff/20496942/)

### CODE REVIEW

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 1]
```diff
@@ -640,6 +640,11 @@ std::shared_ptr<SugaredValue> toSugaredValue(
   }
 
   if (py::isinstance<py::function>(obj)) {
+    if (py::cast<bool>(
+            py::module::import("torch.jit").attr("_is_global_builtin")(obj))) {
+      // See [python globals]
```
Comment (eellison) [inline torch/csrc/jit/python/python_sugared_value.cpp]: Can you add a little bit more than `See [python globals]`, maybe, "don't attempt to compile python builtin" or something  ?

File: torch/csrc/jit/runtime/register_prim_ops.cpp [hunk 2]
```diff
@@ -2727,7 +2727,7 @@ RegisterOperators reg2({
     DEFINE_UNARY_INT_OP(aten::factorial, factorial(a), int),
     DEFINE_UNARY_FLOAT_OP(aten::isnan, std::isnan(a), bool),
     DEFINE_UNARY_FLOAT_OP(aten::isfinite, std::isfinite(a), bool),
-    DEFINE_UNARY_FLOAT_OP(aten::isinf, std::isinf(a), bool),
+    DEFINE_UNARY_FLOAT_OP(aten::isinf, bool(std::isinf(a)), bool),
```
Comment (eellison) [inline torch/csrc/jit/runtime/register_prim_ops.cpp]: accidental revert ?

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 3]
```diff
@@ -487,26 +508,6 @@ struct Environment {
       }
     }
 
-    if (!retval) {
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: Why was this moved ?

File: torch/distributions/von_mises.py [hunk 4]
```diff
@@ -51,7 +51,6 @@ def _log_modified_bessel_fn(x, order=0):
     return result
 
 
-@torch.jit.script
 def _rejection_sample(loc, concentration, proposal_r, x):
```
Comment (eellison) [inline torch/distributions/von_mises.py]: Probably should be removed, i think there was a reason stated in the PR for why this was done.

File: torch/jit/_builtins.py [hunk 5]
```diff
@@ -139,3 +139,49 @@ def _register_builtin(fn, op):
...
 def _find_builtin(fn):
     return _get_builtin_table().get(id(fn))
+
+
+# These are Python globals that have special sugaring in the compiler. See
+# [python globals] in the compiler for details. Adding a builtin global here
+# indicates that when resolving, we should skip these so they dispatch to
+# the correct sugared representation
+compiler_builtins = None
+def _is_global_builtin(maybe_builtin):
```
Comment (eellison) [inline torch/jit/_builtins.py]: maybe is_python_builtin is more accurate

File: torch/jit/_builtins.py [hunk 6]
```diff
@@ -139,3 +139,49 @@ def _register_builtin(fn, op):
...
+    global compiler_builtins
+    if compiler_builtins is None:
+        # Errors (AssertionError and RuntimeError) should not be here since
+        # they are handled separately in the compiler
+        compiler_builtins = set([
```
Comment (eellison) [inline torch/jit/_builtins.py]: Can you replace using https://stackoverflow.com/questions/22864221/is-the-list-of-python-reserved-words-and-builtins-available-in-a-library ? Don't think it's necessary to enumerate them all.

**Review by eellison — COMMENTED** [review]
Looks good, I have a question or two to make sure i'm understanding the change

File: torch/csrc/jit/runtime/register_prim_ops.cpp (hunk 2 above)
Comment (driazati) [reply torch/csrc/jit/runtime/register_prim_ops.cpp]: I'm not sure why (maybe it was my compiler toolchain/libc++, I just switched it), but `isinf` was returning an `int` instead of a `bool` for me. Casting it here should be save either way.

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 3 above)
Comment (driazati) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: For the same reason as values, the lookup needs to happen before we look at the builtins map.

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 7]
```diff
@@ -640,6 +640,13 @@ std::shared_ptr<SugaredValue> toSugaredValue(
   }
 
   if (py::isinstance<py::function>(obj)) {
+    if (py::cast<bool>(
+            py::module::import("torch.jit").attr("_is_python_builtin")(obj))) {
```
[the remark this answers is not in the document]
Comment (driazati) [reply torch/csrc/jit/python/python_sugared_value.cpp]: a `nullptr` for a sugared value is handled and this function is only [used by the resolver](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/script_init.cpp#L92) so it is safe to do. Plus this means what we want it to here, that these values should not get resolved.

File: torch/jit/_builtins.py (hunk 6 above)
Comment (driazati) [reply torch/jit/_builtins.py]: I ran into some problems with this, we support certain builtins as Python calls so they shouldn't be skipped here, we should only skip the ones that we have explicit support for inside the compiler, so we need them to be listed out.
